net: sockets: Fixes net_pkt leak in accept

neja20 opened the pull request:
There is a net_pkt leak in sockets.c which can happen when the tcp stack is under stress and reuses net_context. 

The idea is that when zsock_accepted_cb() is called it installs callback zsock_received_cb which allows tcp stack to start putting net_pkts in the recv_q. However if the connection is closed by a peer when the connection is not yet accepted by the application, the reference count of the net_context is not increased, because zsock_accept_ctx() fails and returns -1 to the application.
Now that when tcp stack calls tcp_conn_unref(), the net_context has only reference count of 1, hence can be reused by another connection, without having the recv_q empty. And finally when the net_context is reused it just k_fifo_init() the recv_q leaking all the net_pkts in recv_q.

My solution simply increases the reference count earlier with instalment of the zsock_received_cb() callback. And consequently flushes recv_q and decrement reference count if zsock_accept_ctx() fails.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -530,6 +539,8 @@ int zsock_accept_ctx(struct net_context *parent, struct sockaddr *addr,
 		if (net_pkt_eof(last_pkt)) {
 			sock_set_eof(ctx);
 			z_free_fd(fd);
+            zsock_flush_queue(ctx);
+            net_context_unref(ctx);
```
There seems to be something wrong with indentation (or could be my browser too).

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -558,19 +571,12 @@ int zsock_accept_ctx(struct net_context *parent, struct sockaddr *addr,
 		} else {
 			z_free_fd(fd);
 			errno = ENOTSUP;
+            zsock_flush_queue(ctx);
+            net_context_unref(ctx);
```
ditto

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -331,6 +331,15 @@ static void zsock_accepted_cb(struct net_context *new_ctx,
 		k_condvar_init(&new_ctx->cond.recv);
```
Please change the commit subject to `net: sockets: Fixes net_pkt leak in accept` and squash the commits to one. Commit msg should also have a body text (description why the patch exists) and signed-off-by line.

neja20 replied:
The most recent PR should resolve all of the mentioned changes.

jukkar reviewed (commented):
Some minor style / conformance issues, the fix itself looks ok.

jukkar commented:
Compliance check script reports some issues
```
Error: See https://docs.zephyrproject.org/latest/contribute/#coding-style for more details.

-:11: WARNING:BLOCK_COMMENT_STYLE: Block comments should align the * on each line
#11: FILE: subsys/net/lib/sockets/sockets.c:336:
+		/* TCP context is effectively owned by both application
+		* and the stack: stack may detect that peer closed/aborted

- total: 0 errors, 1 warnings, 52 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES C99_COMMENT_TOLERANCE COMPLEX_MACRO CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME DT_SCHEMA_BINDING_PATCH DT_SPLIT_BINDING_PATCH ENOSYS FILE_PATH_CHANGES IS_ENABLED_CONFIG MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PREFER_KERNEL_TYPES PREFER_SECTION PRINTK_WITHOUT_KERN_LEVEL REPEATED_WORD SPDX_LICENSE_TAG SPLIT_STRING TRAILING_SEMICOLON UNDOCUMENTED_DT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
Error: See https://docs.zephyrproject.org/latest/contribute/#commit-guidelines for more details

3: UC4 Line exceeds max length (150>75): "Fix net_pkt leak by increasing net_context the reference count earlier in the zsock_accepted_cb() with instalment of the zsock_received_cb() callback."
4: UC4 Line exceeds max length (103>75): "And consequently flushing recv_q and decrement net_context reference count if zsock_accept_ctx() fails."
Error: Process completed with exit code 1.
```

rlubos reviewed (commented):
There's still a problem with the comment (please increase the indentation by one space), the change looks reasonably to me otherwise.

dnejezchleb committed:
net: sockets: Fixes net_pkt leak in accept

Fix net_pkt leak by increasing net_context the reference count earlier
in the zsock_accepted_cb() with instalment of the
zsock_received_cb() callback.

And consequently flushing recv_q and decrement net_context
reference count if zsock_accept_ctx() fails.

Signed-off-by: Daniel Nejezchleb <dnejezchleb@hwg.cz>

neja20 commented:
I am sorry about that coding style, the `checkpath.pl` does not work in my project under Windows.

rlubos reviewed (approved):
LGTM

ycsin commented:
Should this be backported to v2.7?
